### PR TITLE
Display uncovered intermediates by default in item matrix

### PR DESCRIPTION
**Pull request by gcrabbe**
This PR builds on #284, but I figured the additional changes might elicit enough discussion to warrant a separate PR.

At its core, this PR changes the internal plumbing of the `item-matrix` directive to allow rendering uncovered intermediates. At the same time (and this is probably where the discussion comes in), it changes the default behavior to display uncovered intermediates, as I figure it's the more useful alternative (more on that later). An `onlycoveredintermediates` flag, analogous to the existing `onlycovered` and `onlyuncovered` flags, is added to bring back the old behavior (to a point).

The goal of this PR is to support a workflow that I believe should be relatively common: verifying the coverage of requirements that are validated by a testcase who have themselves a test record object attesting to their pass/fail status. Structuring the information in that manner allows to decouple the requirements and testcases (which, as source objects, are version-controlled) from the test records (which, as products of some automated non-regression test suite, are not).

Picture, if you will, the following directive:
```
.. item-matrix:: Testcase traceability matrix
    :source: RQT
    :intermediate: TC
    :target: TR
    :sourcetitle: Requirement
    :intermediatetitle: Testcase
    :targettitle: Test record
    :type: validated_by | tested_by
    :targetcolumns: status
    :splitintermediates:
    :stats:
    :nocaptions:
```

Which would render as:
```
Statistics: 1 out of 1 covered: 100%
+-------------+----------+-------------+--------+
| Requirement | Testcase | Test record | Status |
+-------------+----------+-------------+--------+
| RQT-1       | TC-1     | TR-1        | PASS   |
|             +----------+-------------+--------+
|             | TC-2     | TR-2        | FAIL   |
|             +----------+-------------+--------+
|             | TC-3     | TR-3        | PASS   |
+-------------+----------+-------------+--------+
```

Now, because we'd like to filter the records based on their status without having to parse the table by hand, we can use #284 and modify the directive as follows:
```
.. item-matrix:: Testcase traceability matrix
    :source: RQT
    :intermediate: TC
    :target: TR
    :sourcetitle: Requirement
    :intermediatetitle: Testcase
    :targettitle: Test record
    :type: validated_by | tested_by
    :filtertarget:
    :status: PASS
    :splitintermediates:
    :stats:
    :nocaptions:
```

Producing, before this PR:
```
Statistics: 1 out of 1 covered: 100%
+-------------+----------+-------------+
| Requirement | Testcase | Test record |
+-------------+----------+-------------+
| RQT-1       | TC-1     | TR-1        |
|             +----------+-------------+
|             | TC-3     | TR-3        |
+-------------+----------+-------------+
```

And because we'd also like a failing test record to invalidate the corresponding requirement, we add the `coveredintermediates` flag, resulting in:
```
Statistics: 0 out of 1 covered: 0%
+-------------+----------+-------------+
| Requirement | Testcase | Test record |
+-------------+----------+-------------+
| RQT-1       |          |             |
+-------------+----------+-------------+
```

Giving us correct statistics, but an entirely useless table. In order to know why `RQT-1` is not covered, we would need to split the item matrix into two, mostly negating the point of having an `intermediate` attribute in the first place.

As we can see, `coveredintermediates` (which should probably be renamed to `allintermediatescovered`, `intermediatereductionand`, or something to that effect) actually has two effects, although only the first is documented: requiring all intermediates to be covered in order to cover the source (i.e. reduction AND instead of reduction OR), and hiding all intermediates of an uncovered source. I consider this last effect to be undesirable, which is why I took the liberty of modifying it.

Thus, this PR modifies Table 2 (without the `coveredintermediates` flag) as follows:
```
Statistics: 1 out of 1 covered: 100%
+-------------+----------+-------------+
| Requirement | Testcase | Test record |
+-------------+----------+-------------+
| RQT-1       | TC-1     | TR-1        |
|             +----------+-------------+
|             | TC-2     |             |
|             +----------+-------------+
|             | TC-3     | TR-3        |
+-------------+----------+-------------+
```

And Table 3 (with the `coveredintermediates` flag) as follows:
```
Statistics: 0 out of 1 covered: 0%
+-------------+----------+-------------+
| Requirement | Testcase | Test record |
+-------------+----------+-------------+
| RQT-1       | TC-1     | TR-1        |
|             +----------+-------------+
|             | TC-2     |             |
|             +----------+-------------+
|             | TC-3     | TR-3        |
+-------------+----------+-------------+
```

By adding the `onlycoveredintermediates` flag, Table 2 (without the `coveredintermediates` flag) becomes:
```
Statistics: 1 out of 1 covered: 100%
+-------------+----------+-------------+
| Requirement | Testcase | Test record |
+-------------+----------+-------------+
| RQT-1       | TC-1     | TR-1        |
|             +----------+-------------+
|             | TC-3     | TR-3        |
+-------------+----------+-------------+
```

And Table 3 (with the `coveredintermediates` flag) becomes:
```
Statistics: 0 out of 1 covered: 0%
+-------------+----------+-------------+
| Requirement | Testcase | Test record |
+-------------+----------+-------------+
| RQT-1       | TC-1     | TR-1        |
|             +----------+-------------+
|             | TC-3     | TR-3        |
+-------------+----------+-------------+
```

Which results in a clean split between the coverage (controlled by `coveredintermediates`) and the rendering (controlled by `onlycoveredintermediates`).

### CODE REVIEW

**Comment by JasperCraeghs** [comment]
At first glance, I dislike that the default behavior is changed but the more I think about it, the more I like your approach. The coverage statistics are not affected, are they? 

The current implementation leaves both the intermediate column and target column empty when the source item is not covered. This makes it visually clear which source items are not covered. However, the `:splitintermediates:` option somewhat discards the need to keep the current implementation because it results in at least one empty target cell when the source is not covered.

**Comment by gcrabbe** [comment]
It's not supposed to affect the coverage statistics.

In addition to the MWE in the description, I've verified it's the case on an actual project, and on every corner case I could think of (no intermediates covered, no intermediates at all, etc.).

**Comment by Letme** [comment]
I am a bit against changing the default behavior simply because it produces additional effort for all users. Is there a way to alert users about this default change when upgrading, for example by suggesting at update time to update all matrixes to the old behavior? Maybe the solution is to just define a flag in `conf.py`, something like `append_onlycoveredintermediates_item_matrix = True`?

Secondly, I see with the following statement, we are again limiting flexibility, which in my opinion could be a user-controlled flag of how they want sources covered or not.
```
requiring all intermediates to be covered in order to cover the source (i.e. reduction AND instead of reduction OR), and hiding all intermediates of an uncovered source.
```

File: doc/usage.rst [hunk 1]
```diff
@@ -477,6 +481,11 @@ linked via the ``:intermediate:`` RQT-items:
     When given, an extra column that lists the intermediate item(s) per source item will be added between the columns
     that list sources and the linked targets. The argument will be used as title for this new column.
 
+:onlycoveredintermediates: *optional*, *flag*
+
+    By default, all intermediate items are included. By providing the *onlycoveredintermediates* flag, only covered
```
Comment (JasperCraeghs) [inline doc/usage.rst:486]:
```suggestion
    By default, all intermediate items are shown. By providing this flag, only covered intermediate
```

**Comment by Letme** [comment]
(@JasperCraeghs , any ideas why there is only 1 check on this MR and not the other tests? - I think we have something messed up in GitHub Workflows)

Even if it is a bug (which was not, because it was for a different usecase when it was implemented), it was there for more than 1 year, more than 2 major releases and in the end that means it is the feature users got used to (since nobody proposed a change or complained about it). That means we are bound to break many builds, which is kind of bad. While in release notes we will provide the migration suggestion, I still think we can provide an easy way to get to previous behavior - without having to change a lot of files (and a lot of other tools/plugins which might still produce and rely on the old behavior). For example: Let's assume that `xunit2rst` plugin would produce a matrix for intermediates - that means user updating our traceability extension will have to propose a change in another tool, before they can update the traceability that tool needs to be released and also updated. So with a simple flag in `conf.py` there is a way to produce the old behavior to mitigate the transition.

Secondly: You wrote that now the source is considered covered if *all intermediates* are covered - so using *AND*, but what I wanted to mention that some people out there would use *OR* as acceptable for covering. So that should now also be made configurable if we are re-doing the intermediate behavior.

**Comment by gcrabbe** [comment]
> Even if it is a bug (which was not, because it was for a different usecase when it was implemented), it was there for more than 1 year, more than 2 major releases and in the end that means it is the feature users got used to (since nobody proposed a change or complained about it).

A less generous interpretation would be that nobody complained because nobody tried that particular combination of flags. Even if we take for granted that the current behavior is correct, the fact remains that `coveredintermediates` has two effects, only one of which is documented. Were it a feature in widespread use, this documentation shortcoming would surely have been pointed out by now.

> That means we are bound to break many builds, which is kind of bad.

If by "break many builds" you mean "display slightly more information in a table when a particular combination of flags and conditions is met, without affecting the resulting metrics in any way", then yes.

> While in release notes we will provide the migration suggestion, I still think we can provide an easy way to get to previous behavior - without having to change a lot of files (and a lot of other tools/plugins which might still produce and rely on the old behavior). For example: Let's assume that `xunit2rst` plugin would produce a matrix for intermediates - that means user updating our traceability extension will have to propose a change in another tool, before they can update the traceability that tool needs to be released and also updated. So with a simple flag in `conf.py` there is a way to produce the old behavior to mitigate the transition.

I personally don't like this solution, as it splits the configuration of the `item-matrix` directive across multiple files, making it harder to trace the origin of a given behavior. Moreover, the overriding flag would need to be kept indefinitely because of the same backwards-compatibility argument that justified its introduction. Introducing a long-term increase in (user-facing) complexity to circumvent a short-term (user-facing) discomfort is, to me, the epitome of the cure being worse than the disease.

If, despite my arguments, the proposed change is still deemed unacceptable, an alternative would be to keep `coveredintermediates` as it is, but to deprecate it and introduce `allintermediatesrequired` and `onlycoveredintermediates` as a replacement. `coveredintermediates` would give out a warning when used, preparing users for its removal in a future major release. Since the exact same approach has already been used for deprecating `targetattributes` in favor of `targetcolumns`, it should be an easier pill to swallow.

> Secondly: You wrote that now the source is considered covered if _all intermediates_ are covered - so using _AND_, but what I wanted to mention that some people out there would use _OR_ as acceptable for covering. So that should now also be made configurable if we are re-doing the intermediate behavior.

That's precisely what my PR's `coveredintermediates` does. It switches the coverage criterion from reduction-OR to reduction-AND. If reduction-OR is the desired behavior, simply omit the flag, as it has no other effect.

**Comment by JasperCraeghs** [comment]
> While I would usually agree, I believe the change is warranted in this case because the default behavior is counterproductive. In fact, I would argue it's not so much an intended "default behavior" as an unintended side-effect of how the coveredintermediates feature was first implemented. Before this PR, uncovered intermediates were simply not retrieved, so there was nothing to display.

I assure you that the current behavior is not unintended. I implemented the _intermediate_ option in #200 on  on Jan 5, 2021 so that we can link sources to targets via an intermediate, without even showing the intermediates. Later on, I implemented _intermediatetitle_ in #205 on May 5, 2021, which shows intermediates but only when the source item is covered because otherwise you might have an uncovered source item with the intermediate and target column(s) not empty. This leads to confusion. An uncovered source item should have an empty cell in the target column. The current state of this PR violates this. See the image below, which has the master branch documentation on the left and your current implementation on the right:

Screenshot_20220720_105959

There is no way to render the table on the left with the current state of this PR. This is a problem. Using the *splitintermediates* flag, implemented in #231 on on Jul 14, 2021, will still produce an empty target column but we cannot expect all users to make use of this flag or even be aware of its existence. Merging this PR might lead to confusion for those users.

For this reason, we should maintain backwards compatibility. I agree that *coveredintermediates* needs extra documentation. Please add it. I suggest that you implement two additional flags that replace your *onlycoveredintermediates*:

- *showallintermediates*: renders all intermediate items, even when they are not covered and the source item is not covered.
- *showcoveredintermediates*: renders all covered intermediate items, even when the source item is not covered.

I like that they start with *show*, indicating that they don't affect the filtering/coverage.

Would this suggestion cover all of your use cases?

**Comment by gcrabbe** [comment]
@JasperCraeghs I see. I admit I had not foreseen the impact of this change when `splitintermediates` is not used.

I agree that the following directive:
```
.. item-matrix:: Testcase traceability matrix
    :source: RQT
    :intermediate: TC
    :target: TR
    :sourcetitle: Requirement
    :intermediatetitle: Testcase
    :targettitle: Test record
    :type: validated_by | tested_by
    :filtertarget:
    :status: PASS
    :stats:
    :nocaptions:
```

Should produce (and currently doesn't):
```
Statistics: 1 out of 1 covered: 100%
+-------------+----------+-------------+
| Requirement | Testcase | Test record |
+-------------+----------+-------------+
| RQT-1       | TC-1     | TR-1        |
|             | TC-3     | TR-3        |
+-------------+----------+-------------+
```

And that adding `coveredintermediates` should produce (and currently doesn't):
```
Statistics: 0 out of 1 covered: 0%
+-------------+----------+-------------+
| Requirement | Testcase | Test record |
+-------------+----------+-------------+
| RQT-1       |          |             |
+-------------+----------+-------------+
```
As otherwise the user will get a false impression.

However, I still maintain that `splitintermediates` should produce, in the first case:
```
Statistics: 1 out of 1 covered: 100%
+-------------+----------+-------------+
| Requirement | Testcase | Test record |
+-------------+----------+-------------+
| RQT-1       | TC-1     | TR-1        |
|             +----------+-------------+
|             | TC-2     |             |
|             +----------+-------------+
|             | TC-3     | TR-3        |
+-------------+----------+-------------+
```

And in the second case:
```
Statistics: 0 out of 1 covered: 0%
+-------------+----------+-------------+
| Requirement | Testcase | Test record |
+-------------+----------+-------------+
| RQT-1       | TC-1     | TR-1        |
|             +----------+-------------+
|             | TC-2     |             |
|             +----------+-------------+
|             | TC-3     | TR-3        |
+-------------+----------+-------------+
```
As the current behavior, which is to output a table identical to Table 2 of this comment, is just as misleading as the behavior you decry in my PR. The whole point of `splitintermediates`, and I quote, is to *be useful if you want to group target items per intermediate item instead of per source item*. Hiding `TC-1` and `TC-3` because `TC-2` has no corresponding `TR`, when I specifically asked to make the relationship between each `TC` and `TR` explicit, goes, in my opinion, against that stated goal.

Furthermore, and even though I'm generally a stickler for orthogonality, I'll be the first to recognize that some combinations, such as having both `coveredintermediates` and `onlycoveredintermediates`, make no sense. It's even worse with your proposed `showallintermediates` and `showcoveredintermediates`, which just multiply the number of meaningless configurations (in addition to adopting the opposite convention of every other display control in the extension).

I would argue that the correct approach is to start from the only two use cases that have been put forward so far: if you don't specify `splitintermediates`, it means you're mainly interested in the *source*, and because the merged cell won't allow you to discriminate between covered and uncovered intermediates, they should be hidden altogether; whereas if you do specify `splitintermediates`, it means you're mainly interested in the *intermediates*, and thus they should always be shown, regardless of their coverage status. I believe this would support every known workflow while minimizing the number of configuration knobs, and gently prodding the user away from configurations that might produce misleading outputs.

Thus, I propose to amend this PR in the following way:
* Modify the behavior of `coveredintermediates` flag to take into account the value of `splitintermediates`, as described above.
* Update the description of `coveredintermediates` accordingly.
* Remove the `onlycoveredintermediates` flag.

Therefore, this PR would only affect setups where both `splitintermediates` and `coveredintermediates` are enabled. If you still consider this to be too much of a breaking change, I could preserve the current behavior of `coveredintermediates`, deprecate it, and implement the new behavior through `allintermediatescovered` (which is a better name, anyway). Because `allintermediatescovered` would not be a perfect replacement for `coveredintermediates`, having them both for the next few major releases would allow users to make their voices heard as to why it shouldn't be deprecated.

File: mlx/directives/item_matrix_directive.py [hunk 2]
```diff
@@ -99,17 +100,33 @@ def perform_replacement(self, app, collection):
...
-                covered = source_id in mapping_via_intermediate
-                if covered:
-                    args = [rows, source_item, rights, app]
-                    duplicate_source_count += self._store_source_via_intermediate(mapping_via_intermediate[source_id],
-                                                                                  *args)
+                if not mapping_via_intermediate[source_id]:
+                    self._store_data(rows, source_item, rights, covered, app)
+                    continue
+                covered_intermediates = {intermediate: any(target_set for target_set in target_sets)
+                                         for intermediate, target_sets in mapping_via_intermediate[source_id].items()}
```
Comment (JasperCraeghs) [inline mlx/directives/item_matrix_directive.py:107]:
```suggestion
                                         for intermediate, target_sets in intermediates.items()}
```
This suggestion depends on my first suggestion.

File: mlx/directives/item_matrix_directive.py [hunk 3]
```diff
@@ -99,17 +100,33 @@ def perform_replacement(self, app, collection):
...
+                    for intermediate, target_sets in mapping_via_intermediate[source_id].items():
+                        self._store_row_with_intermediate({intermediate: target_sets},
+                                                            rows, source_item, rights, covered, app)
+                    duplicate_count_total += len(mapping_via_intermediate[source_id]) - 1
+                    duplicate_count_covered += len(mapping_via_intermediate[source_id]) - 1 if covered else 0
+                else:
+                    self._store_row_with_intermediate({intermediate: target_sets for intermediate, target_sets
+                                                        in mapping_via_intermediate[source_id].items()
+                                                        if covered and covered_intermediates[intermediate]},
+                                                        rows, source_item, rights, covered, app)
```
Comment (JasperCraeghs) [inline mlx/directives/item_matrix_directive.py:124]:
```suggestion
                    self._store_row_with_intermediate({intermediate: target_sets for intermediate, target_sets
                                                       in intermediates.items()
                                                       if covered and covered_intermediates[intermediate]},
                                                       rows, source_item, rights, covered, app)
```
This suggestion depends on my first suggestion.

File: mlx/directives/item_matrix_directive.py [hunk 4]
```diff
@@ -99,17 +100,33 @@ def perform_replacement(self, app, collection):
...
+
+                if self['coveredintermediates']:
+                    covered = all(covered_intermediates.values())
+                else:
+                    covered = any(covered_intermediates.values())
+
+                if self['splitintermediates']:
+                    for intermediate, target_sets in mapping_via_intermediate[source_id].items():
+                        self._store_row_with_intermediate({intermediate: target_sets},
+                                                            rows, source_item, rights, covered, app)
```
Comment (JasperCraeghs) [inline mlx/directives/item_matrix_directive.py:117]:
```suggestion
                    for intermediate, target_sets in intermediates.items():
                        self._store_row_with_intermediate({intermediate: target_sets},
                                                           rows, source_item, rights, covered, app)
```
This suggestion depends on my first suggestion.

**Review by JasperCraeghs — CHANGES_REQUESTED** [review]
The functionality seems good. I added some suggestions to make the code more pythonic.

**Review by JasperCraeghs — APPROVED** [review]

**Review by Letme — APPROVED** [review]
Looks fine to me.

**Comment by Letme** [comment]
@gcrabbe Thanks for sticking with our ideas and for making them work with your contribution.

**Comment by JasperCraeghs** [comment]
@gcrabbe Do you have to option to trigger the GitHub workflow that contains the tests?

**Comment by Letme** [comment]
I think we missed pull request when specifying when to run, so #291 should solve it. (then refresh this with master, before it goes in)

File: mlx/directives/item_matrix_directive.py [hunk 5]
```diff
@@ -99,17 +100,35 @@ def perform_replacement(self, app, collection):
...
+
+                if self['coveredintermediates']:
+                    covered = all(covered_intermediates.values())
+                else:
+                    covered = any(covered_intermediates.values())
+
+                if self['splitintermediates']:
+                    for intermediate, target_sets in intermediates.items():
+                        self._store_row_with_intermediate({intermediate: target_sets},
+                                                           rows, source_item, rights, covered, app)
```
Comment (JasperCraeghs) [inline mlx/directives/item_matrix_directive.py:118]:
```suggestion
                                                          rows, source_item, rights, covered, app)
```

File: mlx/directives/item_matrix_directive.py [hunk 6]
```diff
@@ -99,17 +100,35 @@ def perform_replacement(self, app, collection):
...
+                    duplicate_count_for_source = len(intermediates) - 1
+                    duplicate_count_total += duplicate_count_for_source
+                    duplicate_count_covered += duplicate_count_for_source if covered else 0
+                else:
+                    self._store_row_with_intermediate({intermediate: target_sets for intermediate, target_sets
+                                                       in intermediates.items()
+                                                       if covered and covered_intermediates[intermediate]},
+                                                       rows, source_item, rights, covered, app)
```
Comment (JasperCraeghs) [inline mlx/directives/item_matrix_directive.py:126]:
```suggestion
                                                      rows, source_item, rights, covered, app)
```

**Review by JasperCraeghs — CHANGES_REQUESTED** [review]

**Review by JasperCraeghs — APPROVED** [review]

**Pull request merged**